### PR TITLE
Enable setting ServerTimeout and KeepAliveInterval in HubConnectionBuilder

### LINKED ISSUE #18840: Customizable serverTimeoutInMilliseconds in Server-Side Blazor
When debugging a server-side blazor app we often see this error pop up in the client (browser) while stepping through code:

Error: Connection disconnected with error 'Error: Server timeout elapsed without receiving a message from the server.'.

This can be pretty disruptive, requiring a refresh of the browser to recover after continuing the debug session.

Presumably, this happens because the server isn't sending keepalives on the SignalR connection while I'm stepping through code in the debugger. When running in a 'development' environment, I'd like to be able to customize the serverTimeoutInMilliseconds on the HubConnection to prevent this. I see examples of how to customize the SignalR settings for a server-side Blazor app using code like this:

    <environment exclude="Development">
        <script src="_framework/blazor.server.js"></script>
    </environment>
    <environment include="Development">
        <script src="_framework/blazor.server.js" autostart="false"></script>
        <script>
            (function start() {
                Blazor.start({
                    logLevel: 1, // LogLevel.Debug
                    configureSignalR: builder => builder.configureLogging("debug") // LogLevel.Debug
                });
            })();
        </script>
    </environment>

However, I'm unable to change the serverTimeoutInMilliseconds setting. This setting is available on the HubConnection object, but I don't have access to that object. I only have access to HubConnectionBuilder. And HubConnectionBuilder doesn't have a serverTimeoutInMilliseconds property. 

It seems like adding support for configuring the serverTimeoutInMilliseconds via HubConnectionBuilder would be the simplest solution.

Even nicer would be some way to configure this setting in my Startup.cs, probably with a new property available somewhere under the AddServerSideBlazor call, similar to how I can customize the server-side hub settings via AddServerSideBlazor().AddHubOptions(options => ... ).

### LINKED ISSUE #44742: API proposal: enable configuring connection timeouts in HubConnectionBuilder
## Background and Motivation

Related to #18840

When debugging a server-side blazor app customers often see this error pop up in the client (browser) while stepping through code:
Error: Connection disconnected with error 'Error: Server timeout elapsed without receiving a message from the server.'.
This can be disruptive, requiring a refresh of the browser to recover after continuing the debug session.

The workaround for this is to set `serverTimeoutInMilliseconds` for `HubConnection` in `Blazor.start`.

```js
Blazor.start({
    configureSignalR: function (builder) {
        let c = builder.build();
        c.serverTimeoutInMilliseconds = 60000;
        c.keepAliveIntervalInMilliseconds = 30000;
        builder.build = () => {
            return c;
        };
    };
})
```

Instead of overriding build method of `HubConnectionBuilder` with set `serverTimeoutInMilliseconds` it would be nice to set this things directly in the builder:

```js
Blazor.start({
    configureSignalR: function (builder) {
        builder.serverTimeoutInMilliseconds = 60000;
        builder.keepAliveIntervalInMilliseconds = 30000;
    };
});
```

### TypeScript Client

Enable configuring the connection timeouts in `HubConnectionBuilder` like `serverTimeoutInMilliseconds` and `keepAliveIntervalInMilliseconds` .

```diff
class HubConnectionBuilder {
+    public serverTimeoutInMilliseconds?: number;
+
+    public keepAliveIntervalInMilliseconds?: number;

    public reconnectPolicy?: IRetryPolicy;

    public configureLogging(logLevel: LogLevel): HubConnectionBuilder;

    public withUrl(url: string): HubConnectionBuilder;

    public withAutomaticReconnect(): HubConnectionBuilder;
}
```

### C# Client

```diff
public class HubConnectionBuilder : IHubConnectionBuilder
{
    public IServiceCollection Services { get; }

+    public TimeSpan? ServerTimeout { get; set; }

+    public TimeSpan? KeepAliveInterval { get; set; }

    public HubConnectionBuilder();

    public HubConnection Build();
}

public static class HubConnectionBuilderExtensions
{
    public static IHubConnectionBuilder ConfigureLogging(this IHubConnectionBuilder hubConnectionBuilder, Action<ILoggingBuilder> configureLogging);

    public static IHubConnectionBuilder WithAutomaticReconnect(this IHubConnectionBuilder hubConnectionBuilder, TimeSpan[] reconnectDelays);
}

```

## Usage Examples

Customers will use it to configure the connection timeout in server-side blazor app like so:

```js
Blazor.start({
    configureSignalR: function (builder) {
        builder.serverTimeoutInMilliseconds = 60000;
        builder.keepAliveIntervalInMilliseconds = 30000;
    };
});
```

## Alternative Designs

As an alternative we can add a delegate to `Blazor.start` to configure `HubConnection`
```js
Blazor.start({
            configureSignalR: builder => builder.configureLogging("debug"),
            configureConnection: function (connection) {
                connection.serverTimeoutInMilliseconds = 60000;
                connection.keepAliveIntervalInMilliseconds = 30000;
            },
        });
```

## Risks

I'm don't think there are any

### PR DESCRIPTION
**Pull request by surayya-MS**
# Enable setting ServerTimeout and KeepAliveInterval in HubConnectionBuilder

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Added withServerTimeout and withKeepAliveInterval in HubConnectionBuilder for ts and csharp clients

Fixes #18840 and partially #44742

### CODE REVIEW

File: src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs [hunk 1]
```diff
@@ -227,6 +229,21 @@ public HubConnection(IConnectionFactory connectionFactory, IHubProtocol protocol
         _state = new ReconnectingConnectionState(_logger);
 
         _logScope = new ConnectionLogScope();
+
+        var options = serviceProvider.GetService<IOptions<HubConnectionOptions>>();
+
+        if (options != null)
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:235]: nit: We can change this section of code into 2 lines by using nullable operators and the default values
e.g. `ServerTimeout = options?.Value.ServerTimeout ?? DefaultServerTimeout;`

File: src/SignalR/clients/csharp/Client.Core/src/HubConnectionOptions.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.SignalR.Client;
+
+/// <summary>
+/// Configures timeouts for the <see cref="HubConnection" />.
+/// </summary>
+internal class HubConnectionOptions
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client.Core/src/HubConnectionOptions.cs:15]:
```suggestion
internal sealed class HubConnectionOptions
```

File: src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts [hunk 3]
```diff
@@ -183,6 +186,30 @@ export class HubConnectionBuilder {
         return this;
     }
 
+    /** Configures serverTimeoutInMilliseconds for the {@link @microsoft/signalr.HubConnection}.
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts:189]:
```suggestion
    /** Configures {@link @microsoft/signalr.HubConnection.serverTimeoutInMilliseconds} for the {@link @microsoft/signalr.HubConnection}.
```

File: src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts [hunk 4]
```diff
@@ -183,6 +186,30 @@ export class HubConnectionBuilder {
...
+     */
+    public withServerTimeout(milliseconds: number): HubConnectionBuilder{
+        Arg.isRequired(milliseconds, "milliseconds");
+
+        this._serverTimeoutInMilliseconds = milliseconds;
+
+        return this;
+    }
+
+    /** Configures keepAliveIntervalInMilliseconds for the {@link @microsoft/signalr.HubConnection}.
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts:201]:
```suggestion
    /** Configures {@link @microsoft/signalr.HubConnection.keepAliveIntervalInMilliseconds} for the {@link @microsoft/signalr.HubConnection}.
```

File: src/SignalR/clients/csharp/Client/test/UnitTests/HubConnectionBuilderTests.cs [hunk 5]
```diff
@@ -77,4 +77,30 @@ public void AddMessagePackProtocolSetsHubProtocolToMsgPack()
 
         Assert.IsType<MessagePackHubProtocol>(serviceProvider.GetService<IHubProtocol>());
     }
+
+    [Fact]
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client/test/UnitTests/HubConnectionBuilderTests.cs:81]: Can we add a test that configures both (use different values for each)? I want to double check that calling `Configure` on the same options type multiple times works.

File: src/SignalR/clients/csharp/Client.Core/src/HubConnectionBuilderExtensions.cs [hunk 6]
```diff
@@ -63,4 +63,28 @@ public static IHubConnectionBuilder WithAutomaticReconnect(this IHubConnectionBu
...
+        return hubConnectionBuilder;
+    }
+
+    /// <summary>
+    /// Configures KeepAliveInterval for the <see cref="HubConnection" />.
+    /// </summary>
+    /// <param name="hubConnectionBuilder">The <see cref="IHubConnectionBuilder" /> to configure.</param>
+    /// <param name="timeout">KeepAliveInterval for the <see cref="HubConnection"/>.</param>
+    /// <returns>The same instance of the <see cref="IHubConnectionBuilder"/> for chaining.</returns>
+    public static IHubConnectionBuilder WithKeepAliveInterval(this IHubConnectionBuilder hubConnectionBuilder, TimeSpan timeout)
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client.Core/src/HubConnectionBuilderExtensions.cs:85]:
```suggestion
    public static IHubConnectionBuilder WithKeepAliveInterval(this IHubConnectionBuilder hubConnectionBuilder, TimeSpan interval)
```

File: src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts [hunk 7]
```diff
@@ -183,6 +186,30 @@ export class HubConnectionBuilder {
         return this;
     }
 
+    /** Configures {@link @microsoft/signalr.HubConnection.serverTimeoutInMilliseconds} for the {@link @microsoft/signalr.HubConnection}.
+     *
+     * @returns The {@link @microsoft/signalr.HubConnectionBuilder} instance, for chaining.
+     */
+    public withServerTimeout(milliseconds: number): HubConnectionBuilder{
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts:193]:
```suggestion
    public withServerTimeout(milliseconds: number): HubConnectionBuilder {
```

File: src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts [hunk 8]
```diff
@@ -183,6 +186,30 @@ export class HubConnectionBuilder {
...
+        this._serverTimeoutInMilliseconds = milliseconds;
+
+        return this;
+    }
+
+    /** Configures {@link @microsoft/signalr.HubConnection.keepAliveIntervalInMilliseconds} for the {@link @microsoft/signalr.HubConnection}.
+     *
+     * @returns The {@link @microsoft/signalr.HubConnectionBuilder} instance, for chaining.
+     */
+    public withKeepAliveInterval(milliseconds: number): HubConnectionBuilder{
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts:205]:
```suggestion
    public withKeepAliveInterval(milliseconds: number): HubConnectionBuilder {
```

**Review by BrennanConroy — APPROVED** [review]

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts [hunk 9]
```diff
@@ -92,17 +92,38 @@ export class HubConnection {
...
         Arg.isRequired(logger, "logger");
         Arg.isRequired(protocol, "protocol");
 
-        this.serverTimeoutInMilliseconds = DEFAULT_TIMEOUT_IN_MS;
-        this.keepAliveIntervalInMilliseconds = DEFAULT_PING_INTERVAL_IN_MS;
+        if (serverTimeoutInMilliseconds !== undefined) {
+            this.serverTimeoutInMilliseconds = serverTimeoutInMilliseconds;
+        } else {
+            this.serverTimeoutInMilliseconds = DEFAULT_TIMEOUT_IN_MS;
+        }
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/HubConnection.ts:120]: Nit: Is there any reason this cannot use the null coalescing operator? Do we need to treat `null` and undefined differently?

```suggestion
        this.serverTimeoutInMilliseconds = serverTimeoutInMilliseconds ?? DEFAULT_TIMEOUT_IN_MS;
```

**Review by halter73 — APPROVED** [review]
Do we want to do the same thing in HttpHubConnectionBuilder.java? If we do, it doesn't have to be in this PR.

**Comment by surayya-MS** [comment]
> Do we want to do the same thing in HttpHubConnectionBuilder.java? If we do, it doesn't have to be in this PR.

Yes. It will be in another PR.

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts (hunk 9 above)
Comment (surayya-MS) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: Thanks! No particular reason. Your suggestion is better.

**Pull request merged**